### DISCUSSION TITLE
Is there any way to set cluster values other than using `hb_buffer_add`?

### DISCUSSION THREAD

**Discussion by notcome** [harfbuzz/harfbuzz#3161] [Q&A] [answered]
I went through the documentation on `hb-buffer` and it seems that I could only set cluster values by using `hb_buffer_add`. If I want to add code points in batch, those methods (`hb_buffer_add_codepoints`/`hb_buffer_add_utfX`) do not seem to provide a way to customize cluster values. Is there anything I missed?

**Comment by khaledhosny** [comment]
I don’t think you are missing anything. But if you can explain what you want to achieve, there may be a way to do it.

**Reply by notcome** [reply]
I want to have each code point’s cluster value to be its index as grapheme clusters. I am using Swift, and Swift’s String use grapheme clusters as indices. I think sticking with grapheme clusters could save myself lots of trouble.

I understand that even with, say UTF-8 code unit indices, it is still easy to reconstruct cluster values. I am just curious if I could do it at the input level, rather than post-processing Harfbuzz’s output.

**Reply by khaledhosny** [reply]
I don't think indexing by grapheme clusters is going to save you much on the long run as grapheme clusters are too coarse unit (for example, treating a base plus combining mark as one unit might be OK for Latin, but for Arabic it is not that much useful).

**Reply by notcome** [reply]
Yeah… I have just started to play around those complex scripts and I have already felt that. Still, it would be nice to align to Swift’s native String API, to make sure certain types of runtime crash will not happen.

I will just stick with `hb_buffer_add` until my profiler tells me that this becomes a bottleneck. Thanks!

**Comment by behdad** [comment, marked as answer]
The way to set clusters other than what you found is to add your text, then `hb_buffer_get_glyph_infos()` and modify clusters in-place. For example, `hb-shape` does that to use codepoint clustering by default instead of UTF-8 clustering:
https://github.com/harfbuzz/harfbuzz/blob/65c622c6892dcbe44996962a748f13569600d389/util/shape-options.hh#L84-L94
